ENH add agent skill to use benchopt (#959)
benchopt/benchopt: benchopt/skills/benchopt-results/SKILL.md

---
name: benchopt-results
description: "Explore and manage benchopt result files: read run parquets in Python (benchopt.results.read_results), understand the dataframe schema, and use the CLI (plot/merge/publish) plus the outputs/ layout. Use when analysing, slicing, comparing, or sharing benchmark results."
---

# Exploring benchopt results

Every `benchopt run` writes a self-contained result table. You can analyse it
two ways: load the parquet directly in Python (full control, custom analysis)
or use the CLI helpers (`plot`, `merge`, `publish`). Both read the same files.

## Where results live

A run produces, under `<benchmark>/outputs/`:

```
benchopt_run_<timestamp>.parquet   ← the results table (one row per curve point)
benchopt_run_<timestamp>.html      ← interactive dashboard (unless --no-html)
benchopt_run_<timestamp>/          ← per-solver run artifacts (see general.md)
```

`benchopt merge` writes `merged_results.parquet` and `benchopt plot --all`

> writes `all_runs.html` covering every parquet in the folder.

Author: ```suggestion
writes `all_runs.html`, a table of content that allow navigating to HTML reports
for every parquet in the folder.
```

## Read a result file in Python

The public entry point is `read_results` — it handles `.parquet` and `.csv`,
renames the legacy `data_name` column to `dataset_name`, and unpacks any

> objective columns that were pickled at write time:

Author: ```suggestion
objective columns that were pickled at write time (array-api objects):
```

```python
from benchopt.results import read_results

df = read_results("outputs/benchopt_run_2026-06-23_11h08m26.parquet")

> df.shape        # (n_points, n_columns); one row = one sampled point on a curve

Author: ```suggestion
df.shape        # (n_points, n_columns); one row = one sampled point on a curve for a given (dataset, objective, solver, repetition).
```

| Extra metrics | `objective_<name>` | One column per key returned by `Objective.evaluate_result()` (e.g. `objective_train_loss`). |
| Parameters | `p_solver_<param>`, `p_dataset_<param>`, `p_objective_<param>` | One column per parameter, split out of the parametrized name. |
| Provenance | `run_date`, `benchmark-git-tag`, `platform*`, `version-*`, `system-*`, `env-*`, `file_*` | Environment/reproducibility metadata. |

The `name` columns are convenient for grouping; the `p_*` columns are better
for filtering on a single parameter.

> **`time` is training-only.** benchopt pauses the solver timer while
> `Objective.evaluate_result()` runs, for both sampling strategies (callback and
> iteration). So `time` measures only the solver/training cost — evaluation and
> validation are *not* included. Don't read it as end-to-end wall-clock when
> comparing solvers whose evaluation costs differ.

## Slicing common questions

```python
# Final point of each curve (largest stop_val per solver/dataset/repetition)
final = (
    df.sort_values("stop_val")
      .groupby(["dataset_name", "solver_name", "idx_rep"], as_index=False)
      .last()
)

> # Best objective per solver, averaged over repetitions

Author: ```suggestion
# Last objective per solver, averaged over repetitions
```

final.groupby("solver_name")["objective_value"].agg(["mean", "std"])

# Filter on a single parameter instead of the full name
df[df["p_solver_adam_lr"] == 0.0036]
```

To write a (possibly filtered/merged) dataframe back out, use `save_results`,
which defaults to parquet and avoids clobbering by uniquifying the filename:

```python
from benchopt.results import save_results
save_results(final, "outputs/final_points.parquet")
```

> 

Author: ```suggestion

If `final_points.parquet` already exists, the result will be saved as
`final_points-1.parquet`, except if using `uniquify=False`.

```

## CLI: plot, merge, publish

> Run these from the benchmark directory; with no `-f`, they pick the **latest**

Author: ```suggestion
Run these from the benchmark directory; with no `--file/-f`, they pick the **latest**
```
